tests/unit/TagFieldTest.php

```diff
@@ -0,0 +1,395 @@
+	}
+
+	/**
+	 * @test
+	 */
+	function itSuggestsTextTags() {
+		$field = new TagField('TextBasedTags');
+		$field->setRecordClass('TagFieldTest_BlogPost');
+
+		/**
+		 * Partial tag title match.
+		 */
+		$request = new SS_HTTPRequest(
+			'get',
+			'TagFieldTest_Controller/TextBasedTestForm/fields/Tags/suggest',
+			array('term' => 'Text')
+		);
+
+		$this->assertEquals($field->suggest($request), '["Text1","Text2"]');
+
+		/**
+		 * Exact tag title match.
+		 */
+		$request = new SS_HTTPRequest(
+			'get',
+			'TagFieldTest_Controller/TextBasedTestForm/fields/Tags/suggest',
+			array('term' => 'Text1')
+		);
+
+		$this->assertEquals($field->suggest($request), '["Text1"]');
+
+		/**
+		 * Case-insensitive tag title match.
+		 */
+		$request = new SS_HTTPRequest(
+			'get',
+			'TagFieldTest_Controller/TextBasedTestForm/fields/Tags/suggest',
+			array('term' => 'TEXT1')
+		);
+
+		$this->assertEquals($field->suggest($request), '["Text1"]');
+
+		/**
+		 * No tag title match.
+		 */
+		$request = new SS_HTTPRequest(
+			'get',
+			'TagFieldTest_Controller/TextBasedTestForm/fields/Tags/suggest',
+			array('term' => 'unknown')
+		);
+
+		$this->assertEquals($field->suggest($request), '[]');
+	}
+
+	/**
+	 * @test
+	 */
+	function itDisplaysValuesFromRelations() {
+		$form = new Form(
+			$this,
+			'Form',
+			new FieldList(
+				$field = new TagField('Tags')
+			),
+			new FieldList()
+		);
+
+		$form->loadDataFrom(
+			$this->objFromFixture('TagFieldTest_BlogPost', 'BlogPost3')
+		);
+
+		$this->assertEquals($field->Value(), array(1 => 1, 2 => 2));
+	}
+
+	/**
+	 * @test
+	 */
+	function itIgnoresNewTagsIfReadOnly() {
+		$record = new TagFieldTest_BlogPost();
+		$record->write();
+
+		$tag = TagFieldTest_BlogTag::get()->filter('Title', 'Object1')->first();
+
+		$field = new TagField('Tags');
+		$field->setReadOnly(true);
+		$field->setValue(array($tag->ID, 'Object3'));
+		$field->saveInto($record);
+
+		$record = DataObject::get_by_id('TagFieldTest_BlogPost', $record->ID);
+
+		$this->compareExpectedAndActualTags(
+			$record,
+			array('Object1')
+		);
+	}
+}
+
+class TagFieldTest_BlogTag extends DataObject implements TestOnly {
+	/**
+	 * @var string
+	 */
+	private static $default_sort = '"TagFieldTest_BlogTag"."ID" ASC';
+
+	/**
+	 * @var array
+	 */
+	private static $db = array(
+		'Title' => 'Varchar(200)',
+	);
+
+	/**
+	 * @var array
+	 */
+	private static $belongs_many_many = array(
+		'BlogEntries' => 'TagFieldTest_BlogPost',
+	);
+}
+
+class TagFieldTest_BlogPost extends DataObject implements TestOnly {
+	/**
+	 * @var string
+	 */
+	private static $default_sort = '"TagFieldTest_BlogPost"."ID" ASC';
+
+	/**
+	 * @var array
+	 */
+	private static $db = array(
+		'Title' => 'Text',
+		'Content' => 'Text',
+		'TextBasedTags' => 'Text',
+	);
+
+	/**
+	 * @var array
+	 */
+	private static $many_many = array(
+		'Tags' => 'TagFieldTest_BlogTag',
+	);
+}
+
+class TagFieldTest_Controller extends Controller {
```

Comment:
Yes, please. I think `tests/*` classes get excluded from standard manifest builds, but it's still a nice safeguard that's used in core: https://github.com/silverstripe/silverstripe-framework/blob/3.1/tests/control/ControllerTest.php#L430